### PR TITLE
[DO_NOT_MERGE] Investigate rollback_safety scenario trace

### PR DESCRIPTION
**Pull request by heidihoward**
Using the CI to test trace validation and spec.

This PR is not for merging, just for experimenting with

### CODE REVIEW

**Comment by lemmy** [comment]
This change makes TLC still accept `rollback_safety.ndjson` while showing proper action names.

```diff
diff --git a/tla/consensus/Traceccfraft.tla b/tla/consensus/Traceccfraft.tla
index 89ff917e8..f6d7146f6 100644
--- a/tla/consensus/Traceccfraft.tla
+++ b/tla/consensus/Traceccfraft.tla
@@ -373,9 +373,11 @@ TraceNext ==
     \/ IsExecuteAppendEntries
 
     \/ IsRcvProposeVoteRequest
+    \/ /\ l = 191
+       /\ DropMessages \cdot IsRcvRequestVoteRequest
 
 DropAndNext ==
-    \/ IF ENABLED TraceNext THEN TraceNext ELSE DropMessages \cdot TraceNext
+    \/ TraceNext
     \/ IsDropPendingTo
 
 TraceSpec ==
```

File: tests/raft_scenarios/rollback_safety [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+dispatch_one,3
+assert_is_primary,1
+
+# Don't need this heartbeat AE, drop it
+drop_pending_to,1,2
+
+# Now we deliver a stale NACK
+dispatch_single,2,1
+
+# 1 may now believe 0's log is too short!
```
Comment (lemmy) [inline tests/raft_scenarios/rollback_safety:109]: @eddyashton Shouldn't this read "1 may now believe 2's log is too short!" ?  More specifically, from looking at the TLA+ trace, `sentIndex[1][2]=9` changes to `sentIndex[1][2]=4`.

File: tests/raft_scenarios/rollback_safety [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+# Node 0 produces AEs
```
Comment (lemmy) [inline tests/raft_scenarios/rollback_safety:111]: Same, `s/0/1`?

**Review by lemmy — COMMENTED** [review]
This causes `bad_network.ndjson` to fail at state 97.


https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=81461&view=logs&j=0c81e209-1101-513f-1f15-37e9c3575eb2&t=a05664f3-ab91-5eb2-6acd-14e2179ad708

File: tests/raft_scenarios/rollback_safety (hunk 1 above)
Comment (eddyashton) [reply tests/raft_scenarios/rollback_safety:109]: @lemmy - Yes, both this and the next comment had the wrong node IDs. I've corrected that now in #5991 - thanks for spotting!

**Comment by lemmy** [comment]
Emitting two more log lines by appending `dispatch_one,2` to the scenario, and commenting/removing `QuorumLogInv` adds states to the model where node #2's log has been rolled back:

image
 

Question: What non-determinism allows node #2 not to roll back its log at `l=214`?

**Comment by lemmy** [comment]
`main` with 

> Hypothesis why TV doesn't fail:
> 
> 1. Validation in Traceccfraft is generally too weak, i.e., we have to come up with more things to assert about the trace (see @heidihoward _fake.ndjson above)
> 2. The json trace/raft scenario ends prematurely for TV to catch failures (some log lines such as `execute_append_entries_sync` are noops in Traceccfraft).
> 3. `QuorumLogInv` is a conjunct of `TraceMatchesConstraints`, whereas it's commented in e.g. `SIMccfraft.cfg`.
> 4. Occurrence of `\cdot` in `ccfraft`, combined with the incomplete implementation in TLC, may cause problems.
>    4a) The debugger's unsatisfied-next breakpoint and [tlaplus/tlaplus@c2713de](https://github.com/tlaplus/tlaplus/commit/c2713def38aa1d3255a6c2fe2aa037253e00e361) could both be incompatible with \cdot, causing states with a roll-backed log not to be in the model.

4. turned out to be a red herring, and 2. and 3. are optional; 1. (https://github.com/microsoft/CCF/pull/6010) is what makes TV find the violation at state 212 (even in `main`).

**Comment by lemmy** [comment]
> 4. turned out to be a red herring, and 2. and 3. are optional; 1. ([Strengthen assertions in trace validation #6010](https://github.com/microsoft/CCF/pull/6010)) is what makes TV [find the violation](https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=81745&view=logs&j=0c81e209-1101-513f-1f15-37e9c3575eb2&t=a05664f3-ab91-5eb2-6acd-14e2179ad708) at state 212 (even in `main`).  I suggest we continue strengthening TV by working on the various TODOs in `Traceccfraft.tla`.

I take that back for now.

**Comment by lemmy** [comment]
@heidihoward Any reason to keep the `tla-rollback` branch around?

**Comment by heidihoward** [comment]
Nope, fine to delete

**Comment by lemmy** [comment]
Not enough/too weak assertions.

**Pull request closed without merging**